source_code_interpretation/utils/general_py (#31)
Oneflow-Inc/oneflow-yolo-doc: docs/source_code_interpretation/utils/general_py.md

from multiprocessing.pool import ThreadPool  # 多线程模块 线程池
from pathlib import Path  # Path将str转换为Path对象 使字符串路径易于操作的模块
from subprocess import check_output  # 创建一个子进程再命令行执行..., 最后返回执行结果(文件)
from typing import Optional
from zipfile import ZipFile

import cv2 # opencv库
import numpy as np # numpy矩阵处理函数库
import oneflow as flow # OneFlow框架
import oneflow.backends.cudnn as cudnn
import pandas as pd # pandas矩阵操作模块
import pkg_resources as pkg # 用于查找, 自省, 激活和使用已安装的Python发行版
import yaml # yaml配置文件读写模块

from utils.downloads import gsutil_getsize
from utils.metrics import box_iou, fitness

FILE = Path(__file__).resolve()
ROOT = FILE.parents[1]  # YOLOv5 root directory
RANK = int(os.getenv("RANK", -1))

# Settings
DATASETS_DIR = ROOT.parent / "datasets"  # YOLOv5 datasets directory
# 确定最大的线程数 这里被限制在了8
NUM_THREADS = min(8, max(1, os.cpu_count() - 1))  # number of YOLOv5 multiprocessing threads
AUTOINSTALL = str(os.getenv("YOLOv5_AUTOINSTALL", True)).lower() == "true"  # global auto-install mode
VERBOSE = str(os.getenv("YOLOv5_VERBOSE", True)).lower() == "true"  # global verbose mode
FONT = "Arial.ttf"  # https://ultralytics.com/assets/Arial.ttf
# 设置运行相关的一些基本的配置  Settings

> # 控制print打印torch.tensor格式设置  tensor精度为5(小数点后5位)  每行字符数为320个  显示方法为long

Reviewer: ```suggestion
# 控制 print 打印 oneflow.tensor 格式设置  tensor 精度为5(小数点后5位)  每行字符数为320个  显示方法为long
```

flow.set_printoptions(linewidth=320, precision=5, profile="long")
# 控制print打印np.array格式设置  精度为5  每行字符数为320个  format short g, %precision=5
np.set_printoptions(linewidth=320, formatter={"float_kind": "{:11.5g}".format})  # format short g, %precision=5
# pandas的最大显示行数是10
pd.options.display.max_columns = 10
# 阻止opencv参与多线程(与 Pytorch的 Dataloader不兼容)
cv2.setNumThreads(0)  # prevent OpenCV from multithreading (incompatible with PyTorch DataLoader)
os.environ["NUMEXPR_MAX_THREADS"] = str(NUM_THREADS)  # NumExpr max threads

> os.environ["OMP_NUM_THREADS"] = "1" if platform.system() == "darwin" else str(NUM_THREADS)  # OpenMP (Pyflow and SciPy)

Reviewer: ```suggestion
os.environ["OMP_NUM_THREADS"] = "1" if platform.system() == "darwin" else str(NUM_THREADS)  # OpenMP (OneFlow and SciPy)
```

    dealing with wandb login-options timeout issues as well as check_github() timeout issues
    """
    def __init__(self, seconds, *, timeout_msg='', suppress_timeout_errors=True):
        self.seconds = int(seconds)   # 限制时间
        self.timeout_message = timeout_msg  # 报错信息
        self.suppress = bool(suppress_timeout_errors)

    def _timeout_handler(self, signum, frame):
        # 超时处理函数 一旦超时 就在seconds后发送超时信息
        raise TimeoutError(self.timeout_message)

    def __enter__(self):
        # signal.signal: 设置信号处理的函数_timeout_handler
        # 执行流进入with中会执行__enter__方法 如果发生超时, 就会触发超时处理函数_timeout_handler 返回超时异常信息
        signal.signal(signal.SIGALRM, self._timeout_handler)  # Set handler for SIGALRM
        # signal.alarm: 设置发送SIGALRM信号的定时器
        signal.alarm(self.seconds)  # start countdown for SIGALRM to be raised

    def __exit__(self, exc_type, exc_val, exc_tb):
        # 执行流离开 with 块时(没有发生超时), 则调用这个上下文管理器的__exit__方法来清理所使用的资源
        signal.alarm(0)  # Cancel SIGALRM if it's scheduled
        if self.suppress and exc_type is TimeoutError:  # Suppress TimeoutError
            return True

```

## 3.set_logging、init_seeds

这两个函数是一些初始化操作。

> set_logging是对日志的设置(format、level)等进行初始化，init_seeds是进行一系列的随机数种子

Reviewer: ```suggestion
set_logging是对日志的设置(format、level)等进行初始化，init_seeds是设置一系列的随机数种子
```

### 3.1 set_logging

> 这个函数是对日志的格式、等级等进行一个初始化，但是这个函数没用到。

Reviewer: 应该是用到了吧

![图片](https://user-images.githubusercontent.com/35585791/200976842-41dbd6cf-6969-4698-b38b-072f3fc6c61f.png)

```python
def set_logging(name=None, verbose=VERBOSE):
    """广泛使用在train.py、test.py、detect.py等文件的main函数的第一步
    对日志的设置(format、level)等进行初始化
    """
    # Sets level and returns logger
    if is_kaggle():
        for h in logging.root.handlers:
            logging.root.removeHandler(h)  # remove all handlers associated with the root logger object
    rank = int(os.getenv("RANK", -1))  # rank in world for Multi-GPU trainings'
    # 设置日志级别 
    level = logging.INFO if verbose and rank in {-1, 0} else logging.ERROR
    log = logging.getLogger(name)
    log.setLevel(level)
    handler = logging.StreamHandler()
    handler.setFormatter(logging.Formatter("%(message)s"))
    handler.setLevel(level)
    log.addHandler(handler)

```

### 3.2 init_seeds

> 这个函数是使用random.random()、np.random.rand()、init_torch_seeds（调用torch_utils.py中的函数）

Reviewer: ```suggestion
这个函数是使用random.random()、np.random.rand()、init_oneflow_seeds（调用oneflow_utils.py中的函数）
```

等生成一系列的随机数种子，以保证结果的可复现性。

init_seeds函数代码：

```python
def init_seeds(seed=0, deterministic=False):
    # Initialize random number generator (RNG) seeds https://pyflow.org/docs/stable/notes/randomness.html
    # cudnn seed 0 settings are slower and more reproducible, else faster and less reproducible
    # 设置随机数 针对使用random.random()生成随机数的时候相同
    random.seed(seed)    
    # 设置随机数 针对使用np.random.rand()生成随机数的时候相同
    np.random.seed(seed)
    # 为CPU设置种子用于生成随机数的时候相同  并确定训练模式
    flow.manual_seed(seed)
    cudnn.benchmark, cudnn.deterministic = (False, True)
    flow.cuda.manual_seed(seed)
    flow.cuda.manual_seed_all(seed)  # for Multi-GPU, exception safe
```

## 4. get_latest_run

这个函数的作用是查找最近保存的权重文件 last*.pt，用以进行断点续训。

get_latest_run函数代码：

```python
def get_latest_run(search_dir="."):

>     # Return path to most recent 'last.pt' in /runs (i.e. to --resume from)

Reviewer: ```suggestion
    # Return path to most recent 'last' in runs/ (i.e. to --resume from)
```

>     """用在train.py查找最近的pt文件进行断点续训

Reviewer: ```suggestion
    """用在train.py查找最近的oneflow模型文件夹进行断点续训
```

>     用于返回该项目中最近的模型 'last.pt'对应的路径

Reviewer: ```suggestion
    用于返回该项目中最近的模型文件夹 'last'对应的路径
```

    :params search_dir: 要搜索的文件的根目录 默认是 '.'  表示搜索该项目中的文件
    """
    # 从Python版本3.5开始, glob模块支持该"**"指令（仅当传递recursive标志时才会解析该指令)
    # glob.glob函数匹配所有的符合条件的文件, 并将其以list的形式返回
    last_list = glob.glob(f"{search_dir}/**/last", recursive=True)
    # os.path.getctime 返回路径对应文件的创建时间

>     # 所以这里是返回路径列表中创建时间最晚(最近的last文件)的路径

Reviewer: ```suggestion
    # 所以这里是返回路径列表中创建时间最晚(最近的last文件夹)的路径
```

    return max(last_list, key=os.path.getctime) if last_list else ""
```

函数在train.py中被调用：
![image](https://user-images.githubusercontent.com/109639975/199921105-a402c34a-e040-4b8f-b6f3-23f0cad475ef.png)

## 5. is_docker、is_colab、is_pip

> 下面是三个检测函数，is_docker检测当前环境是否是docker环境，

Reviewer: is_pip描述没写完，这三句话可以分3行或者一行写完

这个函数是查询当前环境是否是docker环境，会用到后面的check_git_status和check_imshow等函数中。

is_docker函数代码：

```python
def is_docker() -> bool:
    """
    在后面的check_git_status和check_imshow等函数中被调用
    查询当前环境是否是docker环境  Is environment a Docker container?
    Check if the process runs inside a docker container.
    """
    if Path("/.dockerenv").exists():
        return True
    try:  # check if docker is in control groups
        with open("/proc/self/cgroup") as file:
            return any("docker" in line for line in file)
    except OSError:
        return False
```

### 5.2 is_colab
这个函数是检查当前环境是否是Google Colab环境，会用到后面的check_imshow函数中。

is_colab函数代码：

```python
def is_colab():
    """用到后面的check_imshow函数中
    检查当前环境是否是Google Colab环境  Is environment a Google Colab instance?
    """
    try:
        import google.colab
        return True
    except Exception as e:
        return False
```

### 5.3 is_pip（没用到）
这个函数是检测当前文件是否在pip package(site-packages)文件里，不过这个函数没用到。

is_pip函数代码：

```python
def is_pip():
    """没用到
    当前文件是否在pip package(site-packages)文件里
    Is file in a pip package?
    """
    return 'site-packages' in Path(__file__).absolute().parts
```

## 6. file_size（没用到）

这个函数是返回本地文件的大小，

> 功能和之前google_utils.py中的gsutil_getsize函数（返回网站链接对应文件的大小）很像。

Reviewer: 没有这个文件吧，请给出对应的文件链接和函数位置

            s += f'"{r}" '
            n += 1

    if s and install and AUTOINSTALL:  # check environment variable
        LOGGER.info(f"{prefix} YOLOv5 requirement{'s' * (n > 1)} {s}not found, attempting AutoUpdate...")
        try:
            # 再检查当前主机是否联网
            assert check_online(), "AutoUpdate skipped (offline)"
            # 最后创建一个子进程再执行pip指令并返回执行结果
            LOGGER.info(check_output(f'pip install {s} {cmds}', shell=True).decode())
            source = file if 'file' in locals() else requirements
            s = f"{prefix} {n} package{'s' * (n > 1)} updated per {source}\n" \
                f"{prefix} ⚠️ {colorstr('bold', 'Restart runtime or rerun command for updates to take effect')}\n"
            LOGGER.info(s)
        except Exception as e:
            LOGGER.warning(f'{prefix} ❌ {e}')
```

用在train.py中：

![image](https://user-images.githubusercontent.com/109639975/199922077-d60aa8b2-ac1a-48a6-aeed-57713291825e.png)
val.py 中：

![image](https://user-images.githubusercontent.com/109639975/199922200-7dcdfd99-df58-473c-a9a8-4e47ce981bfd.png)

## 12. make_divisible、check_img_size

> 这两个函数主要是用来约束图像的长款或者feature map的长款，

Reviewer: ```suggestion
这两个函数主要是用来约束图像的长宽或者feature map的长宽，
```

必须是divisor（等于算法的最大下采样率一般是32）的最小倍数。

### 12.1 make_divisible

```python
def make_divisible(x, divisor):

>     """用在下面的make_divisible函数中  yolo.py的parse_model函数和commom.py的AutoShape函数中

Reviewer: 重新组织这句话

    # math.ceil 向上取整
    return math.ceil(x / divisor) * 
```

这个函数用在下面的make_divisible函数中及 yolo.py的parse_model函数和commom.py的AutoShape函数中：

![image](https://user-images.githubusercontent.com/109639975/199922801-16f78242-3973-450d-8f78-880aacd6834f.png)

### 12.2 check_img_size
这个函数是为了保证img_size是能被s（32）整除，如果不能就返回大于等于img_size且是s的最小倍数。

这个函数本质是通过调用make_divisible函数实现的。

check_img_size函数代码：

```python
def check_img_size(imgsz, s=32, floor=0):
    """这个函数主要用于train.py中和detect.py中  用来检查图片的长宽是否符合规定
    检查img_size是否能被s整除，这里默认s为32  返回大于等于img_size且是s的最小倍数
    Verify img_size is a multiple of stride s
    """
    # Verify image size is a multiple of stride s in each dimension
    if isinstance(imgsz, int):  # integer i.e. img_size=640
        # 取大于等于x的最小值且该值能被divisor整除
        new_size = max(make_divisible(imgsz, int(s)), floor)
    else:  # list i.e. img_size=[640, 480]
        imgsz = list(imgsz)  # convert to list if tuple
        new_size = [max(make_divisible(x, int(s)), floor) for x in imgsz]
    if new_size != imgsz:
        LOGGER.warning(f"WARNING: --img-size {imgsz} must be multiple of max stride {s}, updating to {new_size}")
    return new_size
```

> 用来保证img的长宽符合规定，用在val.py , detect.py ,train.py中：

Reviewer: train, val, detect 的顺序吧

![image](https://user-images.githubusercontent.com/109639975/199923004-fc4ff8b8-8647-4092-8fbb-caa7397a30a0.png)

## 13. check_imshow

这个函数是检查一下前环境是否可以使用opencv.imshow显示图片。

```python
def check_imshow():
    """用在detect.py中  使用webcam的时候调用
    检查当前环境是否可以使用opencv.imshow显示图片
    主要有两点限制: Docker环境 + Google Colab环境
    """
    # Check if environment supports image displays
    try:
        assert not is_docker(), "cv2.imshow() is disabled in Docker environments"
        assert not is_colab(), "cv2.imshow() is disabled in Google Colab environments"
        # 初始化一张图片检查下opencv是否可用
        cv2.imshow("test", np.zeros((1, 1, 3)))
        cv2.waitKey(1)
        cv2.destroyAllWindows()
        cv2.waitKey(1)
        return True
    except Exception as e:
        LOGGER.warning(f"WARNING: Environment does not support cv2.imshow() or PIL Image.show() image displays\n{e}")
        return False
```

会在detect.py中使用webcam的时候调用：

![image](https://user-images.githubusercontent.com/109639975/199923396-eff2bcd3-fbca-4755-aa4a-a186a7f5cdcf.png)

## 14. check_file

> 这个函数是检查本都相关文件路径能否找到这个文件，没找到就说明文件丢失了，

Reviewer: ```suggestion
这个函数是检查本地相关文件路径能否找到这个文件，没找到就说明文件丢失了，
```

        else:
            LOGGER.info(f"Downloading {url} to {file}...")
            # 使用flow.hub.download_url_to_file从url地址上中下载文件名为file的文件
            flow.hub.download_url_to_file(url, file)
            # 检查是否下载成功
            assert Path(file).exists() and Path(file).stat().st_size > 0, f"File download failed: {url}"  # check
            # 返回下载的文件名
        return file
    elif file.startswith("clearml://"):  # ClearML Dataset ID
        assert "clearml" in sys.modules, "ClearML is not installed, so cannot use ClearML dataset. Try running 'pip install clearml'."
        return file
    else:  # search
        files = []
        for d in "data", "models", "utils":  # search directories
            files.extend(glob.glob(str(ROOT / d / "**" / file), recursive=True))  # find file
        assert len(files), f"File not found: {file}"  # assert file was found
        assert len(files) == 1, f"Multiple files match '{file}', specify exact path: {files}"  # assert unique
        # 返回第一个匹配到的文件名
        return files[0]  # return file
```

1. 在train.py中使用（检查本地data、cfg、hyp等文件是否存在）
2. 在test.py中使用（检查本地data文件是否存在）

## 15. check_dataset

> 这个函数是检查本地是否有指定的数据集，没用就从torch库中下载并解压数据集。

Reviewer: ```suggestion
这个函数是检查本地是否有指定的数据集，没用就使用oneflow.hub模块在线下载并解压数据集。
```

                flow.hub.download_url_to_file(s, f)
                Path(root).mkdir(parents=True, exist_ok=True)  # create root
                ZipFile(f).extractall(path=root)  # unzip
                Path(f).unlink()  # remove zip
                r = None  # success
            # 如果下载地址s是bash开头就使用bash指令下载数据集
            elif s.startswith("bash "):  # bash script
                LOGGER.info(f"Running {s} ...")
                # 使用bash命令下载
                r = os.system(s)
            # 否则下载地址就是一个python脚本 执行python脚本下载数据集
            else:  # python script
                r = exec(s, {"yaml": data})  # return None
            dt = f"({round(time.time() - t, 1)}s)"
            s = f"success ✅ {dt}, saved to {colorstr('bold', root)}" if r in (0, None) else f"failure {dt} ❌"
            LOGGER.info(f"Dataset download {s}")
    check_font("Arial.ttf" if is_ascii(data["names"]) else "Arial.Unicode.ttf", progress=True)  # download fonts
    return data  # dictionary
```

![image](https://user-images.githubusercontent.com/109639975/199923505-34c41280-15e0-4b7f-8bae-e2f0ecdf4c40.png)

## 16. download
这个函数是将url中的文件下载下来，再解压。但是这个文件并没有在程序中被调用，

> flow.hub.download_url_to_file系统函数和google_utils.py

Reviewer: google_utils.py似乎没有这个文件

中的attempt_download函数进行下载文件。所以，这个函数随便看看就好。

```python
def download(url, dir=".", unzip=True, delete=True, curl=False, threads=1, retry=3):
    # Multi-threaded file download and unzip function, used in data.yaml for autodownload
    """在coco.yaml中下载数据集
    Multi-threaded file download and unzip function
    :params url: 下载文件的url地址
    :params dir: 下载下来文件保存的目录
    :params unzip: 下载后文件是否需要解压
    :params delete: 解压后原文件(未解压)是否需要删除

>     :params curl: 是否使用cmd curl语句下载文件  False就使用torch.hub下载

Reviewer: ```suggestion
    :params curl: 是否使用cmd curl语句下载文件  False就使用oneflow.hub下载
```

            # 解压后是否需要删除未解压的文件
            if delete:
                f.unlink()  # remove zip

    dir = Path(dir)
    dir.mkdir(parents=True, exist_ok=True)  # make directory
    if threads > 1: # 使用线程池
        # 定义了一个线程池, 最多创建threads个线程
        pool = ThreadPool(threads)
        # 进程池中的该方法会将 iterable 参数传入的可迭代对象分成 chunksize 份传递给不同的进程来处理。
        pool.imap(lambda x: download_one(*x), zip(url, repeat(dir)))  # multi-threaded
        pool.close()
        pool.join()
    else:
        for u in [url] if isinstance(url, (str, Path)) else url:
            download_one(u, dir)
```

## 17. clean_str
这个函数是将字符串中一些奇怪的符号 “|@#!¡·$€%&()=?¿^*;:,¨´><+” 换成下划线 ‘_’。

```python
def clean_str(s):
    """在datasets.py中的LoadStreams类中被调用
    字符串s里在pattern中字符替换为下划线_  注意pattern中[]不能省
    Cleans a string by replacing special characters with underscore _
    """
    # re: 用来匹配字符串（动态、模糊）的模块  正则表达式模块
    # pattern: 表示正则中的模式字符串  repl: 就是replacement的字符串  string: 要被处理, 要被替换的那个string字符串
    # 所以这句话执行的是将字符串s里在pattern中的字符串替换为 "_"
    # Cleans a string by replacing special characters with underscore _
    return re.sub(pattern="[|@#!¡·$€%&()=?¿^*;:,¨´><+]", repl="_", string=s)
```

只用在datasets.py中的LoadStreams类中：

> ![image](https://user-images.githubusercontent.com/109639975/199923505-34c41280-15e0-4b7f-8bae-e2f0ecdf4c40.png)

Reviewer: 图贴错了

## 18. one_cycle

这个函数是一种特殊的学习率衰减策略。来自这篇论文： [one_cycle](https://arxiv.org/pdf/1803.09820.pdf). 感兴趣的
朋友可以读一读。

```python
def one_cycle(y1=0.0, y2=1.0, steps=100):
    """用在train.py的学习率衰减策略模块
    one_cycle lr  lr先增加, 再减少, 再以更小的斜率减少
    论文: https://arxiv.org/pdf/1803.09820.pdf
    """
    # lambda function for sinusoidal ramp from y1 to y2 https://arxiv.org/pdf/1812.01187.pdf
    return lambda x: ((1 - math.cos(x * math.pi / steps)) / 2) * (y2 - y1) + y1
```

一般使用one_cycle的效果会比较好。
![image](https://user-images.githubusercontent.com/109639975/199924017-be1893d1-8c23-4624-8fde-ba1d62286346.png)

## 19. labels_to_class_weights & labels_to_image_weights

这两个函数是联合使用的。

最终的目的是为了在数据集中采样的时候，不使用随机采样，而是使用更加科学的按图片权重进行采样。

第一个函数labels_to_class_weights是为了得到数据集中所有类别的权重（频率大的权重小）。

第二个函数labels_to_image_weights是利用labels_to_class_weights函数得到的类别权重得到每张图片对应的一个权重。

然后利用每张图片的权重在当前batch进行采样，这样的采样方式会更加科学点。

两个函数都只在train.py中使用，且是同时使用的如图：
![image](https://user-images.githubusercontent.com/109639975/199924234-f06c87b0-6709-45c6-8f0a-632136a178a0.png)

![image](https://user-images.githubusercontent.com/109639975/199924103-1feb8f29-2ce5-4335-8315-e03e4d6bcf07.png)

## 19.1  labels_to_class_weights

> 这个函数是从训练(gt)标签获得每个类的权重 ，标签频率高的类权重低。

Reviewer: ```suggestion
这个函数是从训练标签(gt)获得每个类的权重 ，标签频率高的类权重低。
```

labels_to_class_weights函数代码：

```python
def labels_to_class_weights(labels, nc=80):
    """用在train.py中  得到每个类别的权重   标签频率高的类权重低
    从训练(gt)标签获得每个类的权重  标签频率高的类权重低
    Get class weights (inverse frequency) from training labels
    :params labels: gt框的所有真实标签labels
    :params nc: 数据集的类别数

>     :return torch.from_numpy(weights): 每一个类别根据labels得到的占比(次数越多权重越小) tensor

Reviewer: ```suggestion
    :return flow.from_numpy(weights): 每一个类别根据labels得到的占比(次数越多权重越小) tensor
```

    """
    if labels[0] is None:  # no labels loaded
        return flow.Tensor()

    labels = np.concatenate(labels, 0)  # labels.shape = (866643, 5) for COCO
    # classes: 所有标签对应的类别labels   labels[:, 0]: 类别   .astype(np.int): 取整
    classes = labels[:, 0].astype(np.int)  # labels = [labels_num, class+xywh]
    # weight: 返回每个类别出现的次数 [1, nc]
    weights = np.bincount(classes, minlength=nc)  # occurrences per class

    # Prepend gridpoint count (for uCE training)
    # gpi = ((320 / 32 * np.array([1, 2, 4])) ** 2 * 3).sum()  # gridpoints per image
    # weights = np.hstack([gpi * len(labels)  - weights.sum() * 9, weights * 9]) ** 0.5  # prepend gridpoints to start

    # 将出现次数为0的类别权重全部取1  replace empty bins with 1
    weights[weights == 0] = 1
    # 其他所有的类别的权重全部取次数的倒数  number of targets per class
    weights = 1 / weights
    # normalize 求出每一类别的占比
    weights /= weights.sum()
    return flow.from_numpy(weights)  # numpy -> tensor

```

## 19.2 labels_to_image_weights

这个函数是利用每张图片真实gt框的真实标签labels和上一步labels_to_class_weights得到的每个类别的权重得到数据集中每

> 张图片对应的权重。

Reviewer: 不需要分行

labels_to_image_weights函数代码：

```python
def labels_to_image_weights(labels, nc=80, class_weights=np.ones(80)):
    """用在train.py中 利用上面得到的每个类别的权重得到每一张图片的权重  再对图片进行按权重进行采样
    通过每张图片真实gt框的真实标签labels和上一步labels_to_class_weights得到的每个类别的权重进行采样
    Produces image weights based on class_weights and image contents
    :params labels: 每张图片真实gt框的真实标签
    :params nc: 数据集的类别数 默认80
    :params class_weights: [80] 上一步labels_to_class_weights得到的每个类别的权重
    """
    # class_counts: 每个类别出现的次数  [num_labels, nc]  每一行是当前这张图片每个类别出现的次数  num_labels=图片数量=label数量
    class_counts = np.array([np.bincount(x[:, 0].astype(np.int), minlength=nc) for x in labels])
    # [80] -> [1, 80]

>     # 整个数据集的每个类别权重[1, 80] *  每张图片的每个类别出现的次数[num_labels, 80] = 得到每一张图片每个类对应的权重[128, 80]

Reviewer: 128应该是num_labels？

    # 另外注意: 这里不是矩阵相乘, 是元素相乘 [1, 80] 和每一行图片的每个类别出现的次数 [1, 80] 分别按元素相乘

>     # 再sum(1): 按行相加  得到最终image_weights: 得到每一张图片对应的采样权重[128]

Reviewer: 这里的128也对应改一下

    return (class_weights.reshape(1, nc) * class_counts).sum(1)
```

## 20. coco80_to_coco91_class
这个函数是将80个类的coco索引换成91类的coco索引。

coco80_to_coco91_class函数代码:

```python
def coco80_to_coco91_class():
    """用在test.py中   从80类映射到91类的coco索引 取得对应的class id
    将80个类的coco索引换成91类的coco索引
    :return x: 为80类的每一类在91类中的位置
    """
    # converts 80-index (val2014) to 91-index (paper)
    # https://tech.amikelive.com/node-718/what-object-categories-labels-are-in-coco-dataset/
    # a = np.loadtxt('data/coco.names', dtype='str', delimiter='\n')
    # b = np.loadtxt('data/coco_paper.names', dtype='str', delimiter='\n')
    # x1 = [list(a[i] == b).index(True) + 1 for i in range(80)]  # darknet to coco
    # x2 = [list(b[i] == a).index(True) if any(b[i] == a) else None for i in range(91)]  # coco to darknet
    x = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 27, 28, 31, 32, 33, 34,
         35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63,
         64, 65, 67, 70, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 84, 85, 86, 87, 88, 89, 90]
    return x

```

> 在val.py中定义：

Reviewer: ```suggestion
在 val.py 中使用：
```

![image](https://user-images.githubusercontent.com/109639975/199926030-7afba57f-21f8-4752-b0ae-952956b16a07.png)

## 21. clip_coords
这个函数的作用是：

将boxes的坐标(x1y1x2y2 左上角右下角)限定在图像的尺寸(img_shape hw)内，防止出界。

这个函数会用在下面的xyxy2xywhn、save_one_boxd等函数中，很重要，必须掌握。

clip_coords函数代码：

```python
def clip_coords(boxes, shape):
    """用在下面的xyxy2xywhn、save_one_boxd等函数中

>     将boxes的坐标(x1y1x2y2 左上角右下角)限定在图像的尺寸(img_shape hw)内

Reviewer: ```suggestion
    将boxes的坐标(x1y1x2y2 左上角右下角)限定在图像的尺寸内
```

    Clip bounding x1y1x2y2 bounding boxes to image shape (height, width)
    """
    # Clip bounding xyxy bounding boxes to image shape (height, width)
    if isinstance(boxes, flow.Tensor):  # faster individually
        boxes[:, 0].clamp_(0, shape[1])  # x1
        boxes[:, 1].clamp_(0, shape[0])  # y1
        boxes[:, 2].clamp_(0, shape[1])  # x2
        boxes[:, 3].clamp_(0, shape[0])  # y2
    else:  # np.array (faster grouped)
        boxes[:, [0, 2]] = boxes[:, [0, 2]].clip(0, shape[1])  # x1, x2
        boxes[:, [1, 3]] = boxes[:, [1, 3]].clip(0, shape[0])  # y1, y2

```

## 22. scale_coords
这个函数是将坐标coords(x1y1x2y2)从img1_shape尺寸缩放到img0_shape尺寸。

x的正坐标是向右，y的正坐标是向下。这个函数也是很重要的。

scale_coords函数代码：

```python
def scale_coords(img1_shape, coords, img0_shape, ratio_pad=None):

>     """用在detect.py和test.py中  将预测坐标从feature map映射回原图

Reviewer: 这里不应该是原图吧，应该是img0？

    :params x: [n, xywh] (x, y):
    :return y: [n, x1y1x2y2] (x1, y1): 左上角  (x2, y2): 右下角
    """
    y = x.clone() if isinstance(x, flow.Tensor) else np.copy(x)
    y[:, 0] = x[:, 0] - x[:, 2] / 2  # top left x
    y[:, 1] = x[:, 1] - x[:, 3] / 2  # top left y
    y[:, 2] = x[:, 0] + x[:, 2] / 2  # bottom right x
    y[:, 3] = x[:, 1] + x[:, 3] / 2  # bottom right y
    return y
```

## 24. xywhn2xyxy & xyxy2xywhn & xyn2xy
这三个函数主要用于dataloaders.py文件中。主要是对图像进行一些变换操作。

xywhn2xyxy是将xywh(normalized) -> x1y1x2y2。xyxy2xywhn是将x1y1x2y2 -> xywh(normalized)。

xyn2xy是将xy(normalized) -> xy。这三个函数也是比较重要的，大家必须掌握。

![image](https://user-images.githubusercontent.com/109639975/199925432-ea7f3142-2800-4de3-96cc-1fcc7acc88b4.png)

### 24.1 xywhn2xyxy

这个函数是xywh(normalized) -> x1y1x2y2。

xywhn2xyxy函数代码:

```python
def xywhn2xyxy(x, w=640, h=640, padw=0, padh=0):
    """用在dataloaders.py的 LoadImagesAndLabels类的__getitem__函数、load_mosaic、load_mosaic9等函数中  
    将xywh(normalized) -> x1y1x2y2   (x, y): 中间点  wh: 宽高   (x1, y1): 左上点  (x2, y2): 右下点
    Convert nx4 boxes from [x, y, w, h] normalized to [x1, y1, x2, y2] where xy1=top-left, xy2=bottom-right
    """

>     y = x.clone() if isinstance(x, torch.Tensor) else np.copy(x)

Reviewer: ```suggestion
    y = x.clone() if isinstance(x, flow.Tensor) else np.copy(x)
```

        # 这样作非极大抑制的时候不同类别的框就不会掺和到一块了  这是一个作nms挺巧妙的技巧
        boxes, scores = x[:, :4] + c, x[:, 4]  # boxes (offset by class), scores
        # 返回nms过滤后的bounding box(boxes)的索引（降序排列）
        # i = torchvision.ops.nms(boxes, scores, iou_thres)  # NMS
        i = flow.nms(boxes, scores, iou_thres)  # NMS
        if i.shape[0] > max_det:  # limit detections
            i = i[:max_det]
        if merge and (1 < n < 3e3):  # Merge NMS (boxes merged using weighted mean)
            # update boxes as boxes(i,4) = weights(i,n) * boxes(n,4)
            iou = box_iou(boxes[i], boxes) > iou_thres  # iou matrix
            weights = iou * scores[None]  # box weights
            # bounding box合并  其实就是把权重和框相乘再除以权重之和
            x[i, :4] = flow.mm(weights, x[:, :4]).float() / weights.sum(1, keepdim=True)  # merged boxes
            if redundant:
                i = i[iou.sum(1) > 1]  # require redundancy

        output[xi] = x[i] # 最终输出  [5, 6]
        # 看下时间超没超时  超时没做完的就不做了
        if (time.time() - t) > time_limit:
            LOGGER.warning(f"WARNING: NMS time limit {time_limit:.3f}s exceeded")
            break  # time limit exceeded

    return output
```

> 这个函数一般会用再detect.py或者val.py的模型前向推理结束之后。

Reviewer: ```suggestion
这个函数一般会用在detect.py或者val.py的模型前向推理结束之后。
```
